### PR TITLE
Using SaltStack to install gluu-flask

### PR DESCRIPTION
**Pull request by quantonganh**
I asked a question in the Skype:

> [4/10/15, 11:34:06 AM] quanta: What is the next steps after creating cluster, ldap, oxauth, oxtrust via Swagger UI?
> [4/10/15, 11:34:26 AM] quanta: How do I dockerize gluuhttpd?
> [4/10/15, 11:34:55 AM] quanta: then I can login to see if the cluster work as expected
> [4/10/15, 11:35:43 AM] quanta: In the /POST node, I just saw this "ldap | oxauth | oxtrust"

### CODE REVIEW

File: saltify.sls [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  pkg:
+    - installed
+    - pkgs:
+      - libssl-dev
+      - python-dev
+      - swig
+      - wget
+  pkgrepo:
+    - managed
+    - name: deb-src http://mirror-fpt-telecom.fpt.net/ubuntu/ {{ grains['oscodename'] }} main restricted universe multiverse
```
Comment (bclermont) [inline saltify.sls]: this is a vietnam ubuntu mirror, you should use global mirror

and why the ``deb-src`` ? it should be ``deb`` right?

File: saltify.sls [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  pkgrepo:
+    - managed
+    - name: deb http://ppa.launchpad.net/saltstack/salt/ubuntu {{ grains['oscodename'] }} main
+    - file: /etc/apt/sources.list.d/saltstack-salt.list
+    - clean_file: True
+    - require_in:
+      - pkg: salt
+  cmd:
+    - run
+    - name: wget -q -O- "http://keyserver.ubuntu.com:11371/pks/lookup?op=get&search=0x4759FA960E27C0A6" | sudo apt-key add -
```
Comment (bclermont) [inline saltify.sls]: is it possible to just use ``ppa``?

but it need ``python-software-properties``

File: saltify.sls [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+git:
+  pkg:
+    - installed
+
+gluu-flask:
+  git:
+    - latest
+    - name: https://github.com/GluuFederation/gluu-flask.git
+    - target: /root/gluu-flask
```
Comment (bclermont) [inline saltify.sls]: is this a good idea to install such application in ``$HOME`` ?

File: saltify.sls [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+python:
+  pkg:
+    - installed
+
+setuptools:
+  file:
+    - managed
+    - name: /usr/local/sbin/ez_setup.py
+    - source: https://bitbucket.org/pypa/setuptools/raw/bootstrap/ez_setup.py
```
Comment (bclermont) [inline saltify.sls]: you must use a specific commit as if the file get updated the ``source_hash`` will change

and also it's not a good idea to always use latest (untested) version

**Comment by bclermont** [comment]
:+1: it probably works now but need little improvement

good!

File: saltify.sls (hunk 4 above)
Comment (bclermont) [inline saltify.sls]: same apply to other usage of directly hosted files in gh and bb

File: saltify.sls (hunk 1 above)
Comment (bclermont) [inline saltify.sls]: oh I see

it's required to rebuild openssl

wonder why :)

File: saltify.sls [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  pkgrepo:
+    - managed
+    - name: deb-src http://archive.ubuntu.com/ubuntu {{ grains['oscodename'] }} main restricted universe multiverse
+    - file: /etc/apt/sources.list.d/gluu-src.list
+    - clean_file: True
+    - require_in:
+      - cmd: flask_depends
+  cmd:
+    - run
+    - name: apt-get -y build-dep openssl
```
Comment (bclermont) [inline saltify.sls]: i guess this require ``pkg: flask_depends``

**Comment by nynymike** [comment]
Looks interesting... definitely different from what I expected. Please continue testing.

**Comment by quantonganh** [comment]
Except the last one:

```
============================= test session starts ==============================
platform linux2 -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
plugins: pythonpath, cov
collected 0 items / 1 errors
Coverage.py warning: No data was collected.

==================================== ERRORS ====================================
______________________________ ERROR collecting . ______________________________
env/local/lib/python2.7/site-packages/py/_path/common.py:334: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
env/local/lib/python2.7/site-packages/py/_path/common.py:380: in gen
    for p in self.gen(subdir):
env/local/lib/python2.7/site-packages/py/_path/common.py:380: in gen
    for p in self.gen(subdir):
env/local/lib/python2.7/site-packages/py/_path/common.py:380: in gen
    for p in self.gen(subdir):
env/local/lib/python2.7/site-packages/py/_path/common.py:380: in gen
    for p in self.gen(subdir):
env/local/lib/python2.7/site-packages/py/_path/common.py:380: in gen
    for p in self.gen(subdir):
env/local/lib/python2.7/site-packages/py/_path/common.py:370: in gen
    if p.check(dir=1) and (rec is None or rec(p))])
env/local/lib/python2.7/site-packages/_pytest/main.py:628: in _recurse
    ihook.pytest_collect_directory(path=path, parent=self)
env/local/lib/python2.7/site-packages/_pytest/main.py:166: in call_matching_hooks
    plugins = self.config._getmatchingplugins(self.fspath)
env/local/lib/python2.7/site-packages/_pytest/config.py:688: in _getmatchingplugins
    plugins += self._conftest.getconftestmodules(fspath)
env/local/lib/python2.7/site-packages/_pytest/config.py:521: in getconftestmodules
    mod = self.importconftest(conftestpath)
env/local/lib/python2.7/site-packages/_pytest/config.py:554: in importconftest
    self._onimport(mod)
env/local/lib/python2.7/site-packages/_pytest/config.py:674: in _onimportconftest
    self.pluginmanager.consider_conftest(conftestmodule)
env/local/lib/python2.7/site-packages/_pytest/core.py:201: in consider_conftest
    if self.register(conftestmodule, name=conftestmodule.__file__):
env/local/lib/python2.7/site-packages/_pytest/core.py:95: in register
    name, plugin, self._name2plugin))
E   ValueError: Plugin already registered: /usr/local/gluu-flask/env/lib/python2.7/site-packages/tests/conftest.pyc=<module 'tests.conftest' from '/usr/local/gluu-flask/env/lib/python2.7/site-packages/tests/conftest.pyc'>
E   {'helpconfig': <module '_pytest.helpconfig' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/helpconfig.pyc'>, 'pytestconfig': <_pytest.config.Config object at 0x7f9056009f10>, 'runner': <module '_pytest.runner' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/runner.pyc'>, 'pythonpath': <module 'pytest_pythonpath' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/pytest_pythonpath.pyc'>, 'unittest': <module '_pytest.unittest' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/unittest.pyc'>, 'pastebin': <module '_pytest.pastebin' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/pastebin.pyc'>, 'skipping': <module '_pytest.skipping' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/skipping.pyc'>, 'genscript': <module '_pytest.genscript' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/genscript.pyc'>, 'session': <Session 'gluu-flask'>, '/usr/local/gluu-flask/tests/conftest.pyc': <module 'conftest' from '/usr/local/gluu-flask/tests/conftest.pyc'>, 'tmpdir': <module '_pytest.tmpdir' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/tmpdir.pyc'>, 'capture': <module '_pytest.capture' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/capture.pyc'>, 'funcmanage': <_pytest.python.FixtureManager instance at 0x7f9055ae32d8>, 'terminalreporter': <_pytest.terminal.TerminalReporter instance at 0x7f9055ada320>, 'assertion': <module '_pytest.assertion' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/assertion/__init__.pyc'>, 'mark': <module '_pytest.mark' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/mark.pyc'>, 'terminal': <module '_pytest.terminal' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/terminal.pyc'>, 'main': <module '_pytest.main' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/main.pyc'>, '140257930548432': <_pytest.config.PytestPluginManager object at 0x7f90582094d0>, 'nose': <module '_pytest.nose' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/nose.pyc'>, 'python': <module '_pytest.python' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/python.pyc'>, '/usr/local/gluu-flask/env/lib/python2.7/site-packages/tests/conftest.pyc': <module 'tests.conftest' from '/usr/local/gluu-flask/env/lib/python2.7/site-packages/tests/conftest.pyc'>, 'recwarn': <module '_pytest.recwarn' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/recwarn.pyc'>, '_cov': <pytest_cov.CovPlugin object at 0x7f9055e08350>, 'monkeypatch': <module '_pytest.monkeypatch' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/monkeypatch.pyc'>, 'resultlog': <module '_pytest.resultlog' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/resultlog.pyc'>, 'cov': <module 'pytest_cov' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/pytest_cov.pyc'>, 'capturemanager': <_pytest.capture.CaptureManager instance at 0x7f9055af0710>, 'junitxml': <module '_pytest.junitxml' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/junitxml.pyc'>, 'doctest': <module '_pytest.doctest' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/doctest.pyc'>, 'pdb': <module '_pytest.pdb' from '/usr/local/gluu-flask/env/local/lib/python2.7/site-packages/_pytest/pdb.pyc'>}
--------------- coverage: platform linux2, python 2.7.6-final-0 ----------------
Name                       Stmts   Miss  Cover   Missing
--------------------------------------------------------
api/__init__                   0      0   100%   
api/app                       23     23     0%   2-37
api/compat                    14     14     0%   2-20
api/database                  52     52     0%   23-100
api/extensions                 4      4     0%   2-8
api/helper/__init__            0      0   100%   
api/helper/common_helper      42     42     0%   23-97
api/helper/docker_helper      89     89     0%   23-191
api/helper/model_helper      108    108     0%   23-232
api/helper/salt_helper        13     13     0%   25-51
api/log                       12     12     0%   23-37
api/model/__init__             4      4     0%   23-26
api/model/base                 5      5     0%   1-10
api/model/gluu_cluster        71     71     0%   22-178
api/model/ldap_node           52     52     0%   23-145
api/model/oxauth_node         29     29     0%   22-73
api/model/oxtrust_node        34     34     0%   22-79
api/reqparser/__init__         2      2     0%   1-2
api/reqparser/cluster         24     24     0%   1-48
api/reqparser/node             4      4     0%   1-8
api/resource/__init__          0      0   100%   
api/resource/cluster          36     36     0%   2-330
api/resource/node             58     58     0%   23-210
api/settings                  28     28     0%   2-43
api/setup/__init__             0      0   100%   
api/setup/base                18     18     0%   23-53
api/setup/ldap_setup         183    183     0%   23-466
api/setup/oxauth_setup       114    114     0%   23-282
api/setup/oxtrust_setup       42     42     0%   23-121
--------------------------------------------------------
TOTAL                       1061   1061     0%   
=========================== 1 error in 4.85 seconds ============================

Summary
-------------
Succeeded: 20 (changed=3)
Failed:     1
-------------
Total states run:     21
```

Everything is OK.

Could you please help me about this?

**Comment by iromli** [comment]
@quantonganh Unfortunately, I can't reproduce the pytest issue above. Still investigating.

**Comment by bclermont** [comment]
Quanta won't be available until Thursday. That would be great if you find the problem before he return!

**Comment by quantonganh** [comment]
@iromli : Could you please please try to run my formula on Trusty (14.04), something like this:

```
salt-call --local state.sls gluu-flask -l debug
```

**Comment by quantonganh** [comment]
I found something: https://github.com/pytest-dev/pytest-django/issues/102

**Comment by quantonganh** [comment]
This is my fault when omit `tests` folder. I have just pushed a fix, re-run, and it's ok now:

```
----------                                                                                                                                           [40/1969]
          ID: gluu-flask
    Function: cmd.wait
        Name: env/bin/py.test tests --cov api --cov-report term-missing
      Result: True
     Comment: Command "env/bin/py.test tests --cov api --cov-report term-missing" run
     Started: 07:30:41.433047
    Duration: 2142.09 ms
     Changes:
              ----------
              pid:
                  11934
              retcode:
                  0
              stderr:

              stdout:
                  ============================= test session starts ==============================
                  platform linux2 -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
                  plugins: pythonpath, cov
                  collected 61 items

                  tests/test_database.py .
                  tests/helper/test_common_helper.py .......
                  tests/helper/test_docker_helper.py ............
                  tests/helper/test_model_helper.py ....
                  tests/helper/test_salt_helper.py ...
                  tests/model/test_gluu_cluster.py .............
                  tests/reqparser/test_reqparser_cluster.py ......
                  tests/resource/test_cluster.py .........
                  tests/resource/test_node.py ......
                  --------------- coverage: platform linux2, python 2.7.6-final-0 ----------------
                  Name                       Stmts   Miss  Cover   Missing
                  --------------------------------------------------------
                  api/__init__                   0      0   100%
                  api/app                       23      0   100%
                  api/compat                    14     14     0%   2-20
                  api/database                  52      0   100%                                                                                      [3/1969]
                  api/extensions                 4      0   100%
                  api/helper/__init__            0      0   100%
                  api/helper/common_helper      43      0   100%
                  api/helper/docker_helper      89      6    93%   94-99
                  api/helper/model_helper       94     30    68%   87-99, 112-116, 122-159
                  api/helper/salt_helper        13      0   100%
                  api/log                       12      0   100%
                  api/model/__init__             4      0   100%
                  api/model/base                 5      0   100%
                  api/model/gluu_cluster        74      0   100%
                  api/model/ldap_node           52      0   100%
                  api/model/oxauth_node         30      0   100%
                  api/model/oxtrust_node        32      0   100%
                  api/reqparser/__init__         2      0   100%
                  api/reqparser/cluster         24      0   100%
                  api/reqparser/node             4      0   100%
                  api/resource/__init__          0      0   100%
                  api/resource/cluster          36      0   100%
                  api/resource/node             75     30    60%   120-125, 128-133, 204-233
                  api/settings                  28      0   100%
                  api/setup/__init__             0      0   100%
                  api/setup/base                39     18    54%   52, 55, 58-59, 62, 65-80
                  api/setup/ldap_setup         215    184    14%   37-51, 54-55, 62-93, 96-162, 165-191, 194-236, 240-314, 318-357, 360-365, 368-426, 429-456,
 463-477, 485-496
                  api/setup/oxauth_setup       116     95    18%   34-104, 107-113, 116-128, 131-152, 160-161, 169-170, 177-178, 182-223, 226-228, 231-238, 24
1-250, 253-258, 261-268, 271-307
                  api/setup/oxtrust_setup       57     45    21%   34-45, 52-59, 69-74, 77-82, 85-105, 108-116, 119-159
                  --------------------------------------------------------
                  TOTAL                       1137    422    63%

                  ========================== 61 passed in 1.92 seconds ===========================

Summary
-------------
Succeeded: 21 (changed=3)
Failed:     0
-------------
Total states run:     21
```

File: salt/weave/init.sls [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  - docker
+
+weave:
+  cmd:
+    - run
+    - name: wget -O /usr/local/bin/weave https://github.com/weaveworks/weave/releases/download/latest_release/weave
+    - require:
+      - pkg: wget
+  module:
+    - wait
```
Comment (bclermont) [inline salt/weave/init.sls]: you can use ``file.managed`` here if you don't use any source it won't replace the content but rather just change the permission.

it's cleaner than ``cmd.run``

**Comment by bclermont** [comment]
I reviewed it, it need just a cosmetic change.

But @quantonganh  tested it today and had been able to use Weave  on a multinode docker setup. and the test had been conclusive.

weave had been kept in a separated formula to be use in other servers. not just glu-flask "master" node.

@nynymike @shouro what you think of this?

**Comment by quantonganh** [comment]
The main thing make me confused is: how weave is used in gluu-flask. If there is any doc, please show me, then I think I can automate this step.

**Pull request closed without merging**